bpo-42021: Fix possible ref leaks during _sqlite3 module init

erlend-aasland opened the pull request:
@corona10, would you mind reviewing this?

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-42021](https://bugs.python.org/issue42021) -->
https://bugs.python.org/issue42021
<!-- /issue-number -->

Commit:
Handle PyDict_SetItemString() errors in _sqlite3

Commit:
Add NEWS

Commit:
Improve naming

corona10 commented on Modules/_sqlite/microprotocols.c:
```diff
@@ -39,12 +39,18 @@ static PyObject *psyco_adapters = NULL;
 int
 pysqlite_microprotocols_init(PyObject *dict)
 {
+    int ret;
```
```suggestion
```

corona10 commented on Modules/_sqlite/microprotocols.c:
```diff
@@ -39,12 +39,18 @@ static PyObject *psyco_adapters = NULL;
     }
 
-    return PyDict_SetItemString(dict, "adapters", psyco_adapters);
+    ret = PyDict_SetItemString(dict, "adapters", psyco_adapters);
```
```suggestion
    int ret = PyDict_SetItemString(dict, "adapters", psyco_adapters);
```

corona10 commented on Modules/_sqlite/module.c:
```diff
@@ -238,12 +238,18 @@ Enable or disable callback functions throwing errors to stderr.");
 
 static void converters_init(PyObject* dict)
 {
+    int ret;
```
```suggestion
```

corona10 commented on Modules/_sqlite/module.c:
```diff
@@ -238,12 +238,18 @@ Enable or disable callback functions throwing errors to stderr.");
 static void converters_init(PyObject* dict)
 {
+    int ret;
+
```
```suggestion
```

corona10 commented on Modules/_sqlite/module.c:
```diff
@@ -238,12 +238,18 @@ Enable or disable callback functions throwing errors to stderr.");
     }
 
-    PyDict_SetItemString(dict, "converters", _pysqlite_converters);
+    ret = PyDict_SetItemString(dict, "converters", _pysqlite_converters);
```
```suggestion
    int ret = PyDict_SetItemString(dict, "converters", _pysqlite_converters);
```

corona10 commented on Modules/_sqlite/module.c:
```diff
@@ -338,6 +344,23 @@ do {                                           \
+    }                                                \
+} while (0)
+
+#define ADD_EXCEPTION(dict, name, exc, base)                    \
```
Since we can pass Module instead of Dict,
I 'd like to suggest to replace to use PyModule_AddObject instead PyDict_SetItemString

corona10 commented on Modules/_sqlite/module.c:
```diff
@@ -338,6 +344,23 @@ do {                                           \
 
+#define ADD_STRING(dict, key, val)                   \
+do {                                                 \
+    if (PyDict_SetItemString(dict, key, val) < 0) {  \
```
Since we can pass Module instead of Dict,
I'd like to suggest to replace to use PyModule_AddObject instead PyDict_SetItemString

corona10 reviewed (changes requested):
Thanks for the work,
I left several reviews for better usage.

The reviews are based on the recommended usage of C API.
We can remove dict and related codes after the review is applied.
```
It is recommended extensions use other PyModule_*() and PyObject_*() functions rather than directly manipulate a module’s __dict__.
```

corona10 commented on Misc/NEWS.d/next/Library/2020-10-12-21-21-24.bpo-42021.8yv_8-.rst:
```diff
@@ -0,0 +1 @@
+Fix possible ref. leaks in :mod:`sqlite3` module init.
```
```suggestion
Fix possible ref leaks in :mod:`sqlite3` module init.
```

corona10 commented on Modules/_sqlite/module.c:
```diff
@@ -338,6 +344,23 @@ do {                                           \
     }                                          \
 } while (0)
 
+#define ADD_STRING(dict, key, val)                   \
```
And ADD_STRING does not look like proper name :)

erlend-aasland replied:
I agree :) Will fix

Commit:
Fix NEWS entry

Co-authored-by: Dong-hee Na <donghee.na92@gmail.com>

Commit:
Address Dong-hee Na's review comments

- Use PyModule_Add*() iso. directly manipulating the module's __dict__
- Fix return type mismatch
- Ditch ADD_STRING (after cleaning up, there was no use for it)

Co-authored-by: Dong-hee Na <donghee.na92@gmail.com>

Commit:
Use PyModule_AddIntMacro() to add SQLite constants

Commit:
Nitpick, declaration parameter naming

corona10 commented on Modules/_sqlite/module.c:
```diff
@@ -338,10 +341,21 @@ do {                                           \
+    if (!exc) {                                                 \
+        goto error;                                             \
+    }                                                           \
+    if (PyModule_AddObject(module, name, exc) < 0) {            \
```
You may need to incref the object before using PyModule_AddObject

https://github.com/python/cpython/blob/037245c5ac46c3436f617a1f5d965929754be239/Modules/_testcapimodule.c#L7156-L7158

corona10 replied:
```
Note

Unlike other functions that steal references, PyModule_AddObject() only decrements the reference count of value on success.

This means that its return value must be checked, and calling code must Py_DECREF() value manually on error. Example usage:
```

erlend-aasland replied:
Yes, I saw that, but I assumed that since `PyErr_NewException` returns a new reference, this was already taken care of. Fixed in bc4d076b05. Thanks!

UPDATE: I somehow misread [C-API/Exceptions](https://docs.python.org/3.10/c-api/exceptions.html?highlight=pyerr_newexception#c.PyErr_NewException) earlier today; seems like `PyErr_NewException` does _not_ return a new reference, which of explains the `Py_INCREF`.

Commit:
INCREF before PyModule_AddObject

Co-authored-by: Dong-hee Na <donghee.na92@gmail.com>

corona10 commented on Modules/_sqlite/module.c:
```diff
@@ -236,14 +236,17 @@ PyDoc_STRVAR(enable_callback_tracebacks_doc,
     if (!_pysqlite_converters) {
         return;
     }
 
```
```suggestion
Py_INCREF(_pysqlite_converters);
```

erlend-aasland replied:
`PyDict_New` returns a new reference, so this should not be necessary, no?

erlend-aasland replied:
Anyway, I've applied the suggested change, but I still don't understand why. From the docs:
```
PyObject *t;

t = PyTuple_New(3);
PyTuple_SetItem(t, 0, PyLong_FromLong(1L));
PyTuple_SetItem(t, 1, PyLong_FromLong(2L));
PyTuple_SetItem(t, 2, PyUnicode_FromString("three"));
```

The above is possible because, as far as I can understand:
1) `PyLong_FromLong` and `PyUnicode_FromString` returns _a new reference_
2) `PyTuple_SetItem` _steals_ the new reference

If `PyTuple_SetItem` _borrowed_ the reference, the above would not be possible.

Further: according to the docs, `PyModule_AddObject` _steals_ a reference when successful, so if we assume (hypothetically) that `PyModule_AddObject` never failed (`Modules/cmathmodule.c` actually uses `PyModule_AddObject` like this!), we should be able to write:
```
PyModule_AddObject(m, "some_number", PyLong_FromLong(...));
PyModule_AddObject(m, "some_exception", PyErr_NewException(...));
PyModule_AddObject(m, "some_dict", PyDict_New(...));
```

However, since `PyModule_AddObject` only steals a reference when successful, we have to clean up if it fails, thus:
```
PyObject *d = PyDict_New(); // returns new reference
int ret = PyModule_AddObject(m, "my_dict", d);
if (ret) { ; } // previous reference was stolen, we're free!
else { Py_DECREF(d); } // reference was not stolen; we have to clean up!
```

Either I completely misunderstand [reference count details](https://docs.python.org/3.10/c-api/intro.html#reference-count-details), or the `*_New()` functions stopped returning new references, or `PyModule_AddObject` does not steal references anymore.

I'd appreciate if you could enlighten me, @corona10 :)

corona10 replied:
@erlend-aasland 
Oh, you are right, my review was wrong.
Can you revert the

```
Py_INCREF(_pysqlite_converters);
Py_INCREF(psyco_adapters);
Py_INCREF(exc);
```

That I pointed out? The latest documents miss `Return value: New reference.` which should be noted and this looks bug

I will create this issue after I check the status.
(So your origin commit was right)

I will check the last review before merging this PR
Thank you :)

erlend-aasland replied:
@corona10 Ah, thanks :) I was re-reading and re-reading the docs, and I just couldn't get it to make sense! I think some of the examples in https://docs.python.org/3.10/extending/extending.html should be fixed as well :) I'll revert the problematic commits and request your review again.

erlend-aasland replied:
Reverted with 5ac7298 and 94e66e1!

`./python.exe -m test -F -r -j1  -R 3:10 test_sqlite` reports success; no leaks

erlend-aasland replied:
PTAL :)

corona10 commented on Modules/_sqlite/microprotocols.c:
```diff
@@ -37,14 +37,19 @@ static PyObject *psyco_adapters = NULL;
     if ((psyco_adapters = PyDict_New()) == NULL) {
         return -1;
     }
 
```
```suggestion
Py_INCREF(psyco_adapters);
```

Commit:
Use Py_INCREF after PyDict_New

Co-authored-by: Dong-hee Na <donghee.na92@gmail.com>

Commit:
Revert "Use Py_INCREF after PyDict_New"

This reverts commit 6a5a7cbb6c653cba9b41e96ef3f999707ba0fcf4.

Commit:
Revert "INCREF before PyModule_AddObject"

This reverts commit bc4d076b058d523065b10d1f3e1f6346de15f38c.

corona10 commented on Modules/_sqlite/module.c:
```diff
@@ -264,59 +267,52 @@ static PyMethodDef module_methods[] = {
-};
-
+    ret += PyModule_AddIntMacro(module, SQLITE_DONE);
+    return ret;
```
@erlend-aasland 
This code does not return fail if the PyModule_AddIntMacro is failed partially.
But the original code was goto error even in the single failure.

So I'd like to change this code return 0 if the all code is success and return -1 on single failure.

```c
static int add_integer_constants(PyObject *module) {
      if (PyModule_AddIntMacro(module, PARSE_DECLTYPES) < 0) {
                  return -1;
      }
      return 0;
}
```

corona10 replied:
origin code
```c
   for (i = 0; _int_constants[i].constant_name != NULL; i++) {
        tmp_obj = PyLong_FromLong(_int_constants[i].constant_value);
        if (!tmp_obj) {
            goto error;
        }
        PyDict_SetItemString(dict, _int_constants[i].constant_name, tmp_obj);
        Py_DECREF(tmp_obj);
    }
```

I know this is a slightly different behavior because it only checks the creation of tmp_obj is failed or not
But IMHO PyModule_AddIntMacro is the same as PyLong_FromLong + PyDict_SetItemString so we should handle partial failure.

erlend-aasland replied:
Hm, `PyModule_AddIntMacro` returns 0 or -1. For each macro, I add the return value to `ret`, so if there's no failures, `ret` is 0, and if there's one or more failures, `ret` is negative (for example -2 if there was two failures.) This should not be a problem, but I can adjust it to return immediately on the first failure if that's preferred.

erlend-aasland replied:
> I know this is a slightly different behavior because it only checks the creation of tmp_obj is failed or not
> But IMHO PyModule_AddIntMacro is the same as PyLong_FromLong + PyDict_SetItemString so we should handle partial failure.

No problem, I'll fix it. I appreciate your thorough review!

corona10 replied:
Ah okay,

corona10 replied:
Let's leave this code as you did :)

erlend-aasland replied:
Great, ok :)

erlend-aasland replied:
Nitpick: I'd add a single blank line before `return ret` to be more in sync with PEP 7.

corona10 reviewed (approved):
lgtm
